libfreefare/libfreefare/mifare_desfire_key.3
.\" Copyright (C) 2010 Romain Tartiere
.\"
.\" This program is free software: you can redistribute it and/or modify it
.\" under the terms of the GNU Lesser General Public License as published by the
.\" Free Software Foundation, either version 3 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
.\" FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
.\" more details.
.\"
.\" You should have received a copy of the GNU Lesser General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>
.\"
.\" $Id: mifare_desfire_key.3 730 2010-12-24 13:34:25Z rtartiere@il4p.fr $
.\"
.Dd July 20, 2010
.Dt MIFARE_DESFIRE_KEY 3
.Os
.\"  _   _
.\" | \ | | __ _ _ __ ___   ___
.\" |  \| |/ _` | '_ ` _ \ / _ \
.\" | |\  | (_| | | | | | |  __/
.\" |_| \_|\__,_|_| |_| |_|\___|
.\"
.Sh NAME
.Nm mifare_desfire_des_key_new ,
.Nm mifare_desfire_3des_key_new ,
.Nm mifare_desfire_3k3des_key_new ,
.Nm mifare_desfire_aes_key_new ,
.Nm mifare_desfire_des_key_new_with_version ,
.Nm mifare_desfire_3des_key_new_with_version ,
.Nm mifare_desfire_3k3des_key_new_with_version ,
.Nm mifare_desfire_aes_key_new _with_version,
.Nm mifare_desfire_key_get_version ,
.Nm mifare_desfire_key_set_version ,
.Nm mifare_desfire_key_free
.Nd Mifare DESFire keys Manipulation Functions
.\"  _     _ _
.\" | |   (_) |__  _ __ __ _ _ __ _   _
.\" | |   | | '_ \| '__/ _` | '__| | | |
.\" | |___| | |_) | | | (_| | |  | |_| |
.\" |_____|_|_.__/|_|  \__,_|_|   \__, |
.\"                               |___/
.Sh LIBRARY
Mifare card manipulation library (libfreefare, \-lfreefare)
.\"  ____                              _
.\" / ___| _   _ _ __   ___  _ __  ___(_)___
.\" \___ \| | | | '_ \ / _ \| '_ \/ __| / __|
.\"  ___) | |_| | | | | (_) | |_) \__ \ \__ \
.\" |____/ \__, |_| |_|\___/| .__/|___/_|___/
.\"        |___/            |_|
.Sh SYNOPSIS
.In freefare.h
.Ft MifareDESFireKey
.Fn mifare_desfire_des_key_new "uint8_t value[8]"
.Ft MifareDESFireKey
.Fn mifare_desfire_3des_key_new "uint8_t value[16]"
.Ft MifareDESFireKey
.Fn mifare_desfire_3k3des_key_new "uint8_t value[24]"
.Ft MifareDESFireKey
.Fn mifare_desfire_aes_key_new "uint8_t value[16]"
.Ft MifareDESFireKey
.Fn mifare_desfire_des_key_new_with_version "uint8_t value[8]"
.Ft MifareDESFireKey
.Fn mifare_desfire_3des_key_new_with_version "uint8_t value[16]"
.Ft MifareDESFireKey
.Fn mifare_desfire_3k3des_key_new_with_version "uint8_t value[24]"
.Ft MifareDESFireKey
.Fn mifare_desfire_aes_key_new_with_version "uint8_t value[16]" "uint8_t version"
.Ft uint8_t
.Fn mifare_desfire_key_get_version "MifareDESFireKey key"
.Ft void
.Fn mifare_desfire_key_set_version "MifareDESFireKey key" "uint8_t version"
.Ft void
.Fn mifare_desfire_key_free "MifareDESFireKey key"
.\"  ____                      _       _   _
.\" |  _ \  ___  ___  ___ _ __(_)_ __ | |_(_) ___  _ __
.\" | | | |/ _ \/ __|/ __| '__| | '_ \| __| |/ _ \| '_ \
.\" | |_| |  __/\__ \ (__| |  | | |_) | |_| | (_) | | | |
.\" |____/ \___||___/\___|_|  |_| .__/ \__|_|\___/|_| |_|
.\"                             |_|
.Sh DESCRIPTION
The
.Fn mifare_desfire_key_*
family of functions allows management of Mifare DESFire keys.
.Pp
The
.Fn mifare_desfire_des_key_new ,
.Fn mifare_desfire_3des_key_new ,
.Fn mifare_desfire_3k3des_key_new
and
.Fn mifare_desfire_ades_key_new
alocate a new key with the provided data
.Va value .
The key version is set to
.Va 0 .
.Pp
The
.Fn mifare_desfire_des_key_new_with_version ,
.Fn mifare_desfire_3des_key_new_with_version ,
.Fn mifare_desfire_3k3des_key_new_with_version
and
.Fn mifare_desfire_aes_key_new_with_version
functions are equivalent to the previous functions except that the key version
is read from
.Va value
(DES, 3DES, 3K3DES) or set to
.Va version
(AES).
.Pp
The version of a
.Vt MifareDESFireKey
can be extracted using
.Fn mifare_desfire_key_get_version
and changed using
.Fn mifare_desfire_key_set_version.
.Pp
The
.Fn mifare_desfire_key_free
has to be called for each
.Va MifareDESFireKey
after usage to reclaim memory.
.\"  ____      _                                 _
.\" |  _ \ ___| |_ _   _ _ __ _ __   __   ____ _| |_   _  ___  ___
.\" | |_) / _ \ __| | | | '__| '_ \  \ \ / / _` | | | | |/ _ \/ __|
.\" |  _ <  __/ |_| |_| | |  | | | |  \ V / (_| | | |_| |  __/\__ \
.\" |_| \_\___|\__|\__,_|_|  |_| |_|   \_/ \__,_|_|\__,_|\___||___/
.\"
.Sh RETURN VALUES
Key allocations functions return the allocaed key of
.Va NULL
on failure.
.\"  ____                    _
.\" / ___|  ___  ___    __ _| |___  ___
.\" \___ \ / _ \/ _ \  / _` | / __|/ _ \
.\"  ___) |  __/  __/ | (_| | \__ \ (_) |
.\" |____/ \___|\___|  \__,_|_|___/\___/
.\"
.Sh SEE ALSO
.Xr mifare_desfire 3
.\"     _         _   _
.\"    / \  _   _| |_| |__   ___  _ __ ___
.\"   / _ \| | | | __| '_ \ / _ \| '__/ __|
.\"  / ___ \ |_| | |_| | | | (_) | |  \__ \
.\" /_/   \_\__,_|\__|_| |_|\___/|_|  |___/
.\"
.Sh AUTHORS
.An Romain Tartiere Aq romain@il4p.org


libfreefare-0.3.1/ChangeLog
# Generated by Makefile. Do not edit.

2011-02-23 15:13  Romain Tartière <rtartiere@il4p.fr>

	* TAG libfreefare-0.3.1.

2011-02-23 15:12  Romain Tartière <rtartiere@il4p.fr>

	* NEWS, configure.ac: libfreefare: Update for 0.3.1.

2011-02-23 15:08  Romain Tartière <rtartiere@il4p.fr>

	* libfreefare/mifare_desfire_crypto.c: libfreefare: R is const.

2011-02-13 14:26  Romain Tartière <rtartiere@il4p.fr>

	* test/Makefile.am, test/common/Makefile.am: libfreefare: Do not
	  systematically build the tests if cutter is installed.
	  
	  Building them only when running the test-suite is probably
	  enough.

2011-02-13 14:20  Romain Tartière <rtartiere@il4p.fr>

	* Makefile.am: libfreefare: Shorter is better.

2011-02-13 14:13  Romain Tartière <rtartiere@il4p.fr>

	* NEWS, libfreefare/Makefile.am, libfreefare/freefare.h,
	  libfreefare/freefare_error.3, libfreefare/mifare_desfire.c,
	  libfreefare/mifare_desfire_crypto.c,
	  libfreefare/mifare_desfire_error.c, test/test_mifare_desfire.c:
	  libfreefare: Do not abort on crypto error.
	  
	  Because in some circumstances the crypto is skipped regardless of
	  the
	  communication settings by the Mifare DESFire (e.g. when reading a
	  file which is
	  writable with any key), do not abort if the crypto fail, and make
	  it possible
	  to the user to catch such an event to fix his code accordingly.
	  
	  Only display crypto diagnostic if compiled with debug support.

2011-02-13 12:52  Romain Tartière <rtartiere@il4p.fr>

	* Makefile.am: libfreefare: prevent myself from shooting in my own
	  foot.

2011-02-13 12:50  Romain Tartière <rtartiere@il4p.fr>

	* NEWS, libfreefare/mifare_classic.c: libfreefare: Fix
	  mifare_classic_transfer() with some readers.

2011-01-01 15:27  Romain Tartière <rtartiere@il4p.fr>

	* libfreefare/mifare_desfire_crypto.c: libfreefare: Fix CRC
	  localisation code.
	  
	  New issue
	  Summary: Add regression tests for CRC location in encyphered data
	  files.
	  The CRC position location code in mifare_cryto_postprocess_data()
	  shall be
	  checked.
	  
	  new issue
	  Summary: Rework mifare_cryto_postprocess_data() CRC localisations
	  The function has two different implementation of the same
	  feature: locate a CRC
	  at the end of a decyphered stream and check it. The crc32 will
	  fail if the
	  last CRC byte is 0x80, and the crc16 code looks awful. A
	  refactoring would
	  make our life easier.

2011-01-01 12:46  Romain Tartière <rtartiere@il4p.fr>

	* examples/mifare-desfire-info.c: libfreefare: Do not exit if the
	  master key settings are not readable.

2010-12-31 12:47  Romain Tartière <rtartiere@il4p.fr>

	* libfreefare/mifare_desfire.c: libfreefare: Make
	  mifare_desfire_authenticate() a bit more magic.
	  
	  If the provided key is a 3K3DES key, authenticate in ISO mode. If
	  the key is
	  an AES one, authenticate in AES mode. This sugar should help
	  third party
	  applications to handle authentication in a more generic way.

2010-12-31 10:53  Romain Tartière <rtartiere@il4p.fr>

	* libfreefare/mifare_desfire.c: libfreefare: Move away global
	  variables.

2010-12-29 22:18  Romain Tartière <rtartiere@il4p.fr>

	* libfreefare/mifare_desfire.c: libfreefare: Fix create_file2().
	  
	  ECOPYPASTETOOFAST | EBADLUCKITSVALID: The argument
	  communication_settings is
	  not to be used to construct the parameters passed to the crypto
	  functions
	  mifare_cryto_preprocess_data() & mifare_cryto_postprocess_data(),
	  but only for
	  building the frame.

2010-12-29 22:05  Romain Tartière <rtartiere@il4p.fr>

	* libfreefare/mifare_desfire.c, test/test_mifare_desfire_ev1_aes.c:
	  libfreefare: Fix write_data() return value when data is
	  enciphered.
	  
	  We only have to return the number of bytes of raw data which are
	  sent.

2010-12-29 21:44  Romain Tartière <rtartiere@il4p.fr>

	* libfreefare/mifare_desfire.c: libfreefare: Fix usage of a
	  potentially reallocated memory area.

2010-12-28 11:42  Romain Tartière <rtartiere@il4p.fr>

	* libfreefare/freefare.c: libfreefare: Fix parenthesis position.

2010-12-28 11:30  Romain Tartière <rtartiere@il4p.fr>

	* ., libfreefare/Makefile.am, libfreefare/freefare.c,
	  libfreefare/freefare.h, libfreefare/freefare_internal.h,
	  libfreefare/mifare_desfire_crypto.c,
	  libfreefare/mifare_ultralight.c,
	  test/mifare_ultralight_fixture.c, test/test_mifare_ultralight.c:
	  libfreefare: Merge de libfreefare-ultralight branch into trunk
	  (Mifare UltraLightC support).

2010-12-26 15:46  Romain Tartière <rtartiere@il4p.fr>

	* NEWS, libfreefare/Makefile.am, libfreefare/mifare_desfire.3:
	  libfreefare: Update man pages.

2010-12-25 09:46  Romain Tartière <rtartiere@il4p.fr>

	* examples, examples/mifare-desfire-ev1-configure-ats.c:
	  libfreefare: Set the default ATS instead of a custom one.
	  
	  This prevent some tools to work properly with a card afterthat.

2010-12-24 22:14  Romain Tartière <rtartiere@il4p.fr>

	* libfreefare/freefare.c, libfreefare/freefare_internal.h:
	  libfreefare: Detect Mifare DESFire targets with more than one
	  historical character.

2010-12-24 22:08  Romain Tartière <rtartiere@il4p.fr>

	* NEWS, examples/Makefile.am,
	  examples/mifare-desfire-ev1-configure-ats.c,
	  libfreefare/freefare.h, libfreefare/mifare_desfire.c:
	  libfreefare: New API function mifare_desfire_set_ats().

2010-12-24 22:08  Romain Tartière <rtartiere@il4p.fr>

	* libfreefare/mifare_desfire_crypto.c: libfreefare: Simplify
	  redundant code.

2010-12-24 21:10  Romain Tartière <rtartiere@il4p.fr>

	* examples/mifare-desfire-ev1-configure-default-key.c: libfreefare:
	  Fix example.

2010-12-24 20:41  Romain Tartière <rtartiere@il4p.fr>

	* NEWS, examples, examples/Makefile.am,
	  examples/mifare-desfire-ev1-configure-default-key.c,
	  libfreefare/freefare.h, libfreefare/mifare_desfire.c:
	  libfreefare: New API function mifare_desfire_set_default_key().

2010-12-24 20:39  Romain Tartière <rtartiere@il4p.fr>

	* test/test_mifare_desfire_des.c: libfreefare: Commit files

nfc-tools-read-only/libfreefare/mifare_desfire_key.c
/*-
 * Copyright (C) 2010, Romain Tartiere.
 * 
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 * 
 * $Id: mifare_desfire_key.c 568 2010-09-03 18:04:12Z rtartiere@il4p.fr $
 */

#include <stdlib.h>
#include <string.h>

#include <openssl/des.h>

#include <freefare.h>
#include "freefare_internal.h"

static inline void update_key_schedules (MifareDESFireKey key);

static inline void
update_key_schedules (MifareDESFireKey key)
{
    DES_set_key ((DES_cblock *)key->data, &(key->ks1));
    DES_set_key ((DES_cblock *)(key->data + 8), &(key->ks2));
}

MifareDESFireKey
mifare_desfire_des_key_new (uint8_t value[8])
{
    uint8_t data[8];
    memcpy (data, value, 8);
    for (int n=0; n < 8; n++)
	data[n] &= 0xfe;
    return mifare_desfire_des_key_new_with_version (data);
}

MifareDESFireKey
mifare_desfire_des_key_new_with_version (uint8_t value[8])
{
    MifareDESFireKey key;

    if ((key = malloc (sizeof (struct mifare_desfire_key)))) {
	memcpy (key->data, value, 8);
	memcpy (key->data+8, value, 8);
	update_key_schedules (key);
	key->type = T_DES;
    }
    return key;
}

MifareDESFireKey
mifare_desfire_3des_key_new (uint8_t value[16])
{
    uint8_t data[16];
    memcpy (data, value, 16);
    for (int n=0; n < 8; n++)
	data[n] &= 0xfe;
    for (int n=8; n < 16; n++)
	data[n] |= 0x01;
    return mifare_desfire_3des_key_new_with_version (data);
}

MifareDESFireKey
mifare_desfire_3des_key_new_with_version (uint8_t value[16])
{
    MifareDESFireKey key;

    if ((key = malloc (sizeof (struct mifare_desfire_key)))) {
	memcpy (key->data, value, 16);
	update_key_schedules (key);
	key->type = T_3DES;
    }
    return key;
}

uint8_t
mifare_desfire_key_get_version (MifareDESFireKey key)
{
    uint8_t version = 0;

    for (int n = 0; n < 8; n++) {
	version |= ((key->data[n] & 1) << (7 - n));
    }

    return version;
}

void
mifare_desfire_key_set_version (MifareDESFireKey key, uint8_t version)
{
    for (int n = 0; n < 8; n++) {
	uint8_t version_bit = ((version & (1 << (7-n))) >> (7-n));
	key->data[n] &= 0xfe;
	key->data[n] |= version_bit;
	if (key->type == T_DES) {
	    key->data[n+8] = key->data[n];
	} else {
	    // Write ~version to avoid turning a 3DES key into a DES key
	    key->data[n+8] &= 0xfe;
	    key->data[n+8] |= ~version_bit;
	}
    }
}

MifareDESFireKey
mifare_desfire_session_key_new (uint8_t rnda[8], uint8_t rndb[8], MifareDESFireKey authentication_key)
{
    MifareDESFireKey key;

    uint8_t buffer[16];
    memcpy (buffer, rnda, 4);
    memcpy (buffer+4, rndb, 4);
    memcpy (buffer+8, rnda+4, 4);
    memcpy (buffer+12, rndb+4, 4);

    switch (authentication_key->type) {
    case T_DES:
	key = mifare_desfire_des_key_new_with_version (buffer);
	break;
    case T_3DES:
	key = mifare_desfire_3des_key_new_with_version (buffer);
	break;
    }

    return key;
}

void
mifare_desfire_key_free (MifareDESFireKey key)
{
    free (key);
}
